internal/provider/resource_workspace_deployed_module.go

```diff
@@ -0,0 +1,469 @@
+		diags.AddError(msg, "")
+		return diags
+	}
+
+	if err = t.waitForJobCompletion(ctx, appliedRun.Apply.CurrentJobID); err != nil {
+		diags.AddError("Failed to wait for apply job completion", err.Error())
+		return diags
+	}
+
+	finishedRun, err := t.client.Run.GetRun(ctx, &sdktypes.GetRunInput{ID: appliedRun.Metadata.ID})
+	if err != nil {
+		diags.AddError("Failed to get finished run", err.Error())
+		return diags
+	}
+
+	// If an apply job succeeds, finishedRun.Status is "applied" and
+	// finishedRun.Apply.Status is "finished".
+	if finishedRun.Status != "applied" {
+		diags.AddError("Apply failed", string(finishedRun.Status))
+		return diags
+	}
+	if finishedRun.Apply.Status != "finished" {
+		diags.AddError("Apply status", string(finishedRun.Apply.Status))
+		return diags
+	}
+
+	// In case of a rainy day, make sure the ModuleSource and ModuleVersion *string aren't nil.
+	if finishedRun.ModuleSource == nil {
+		diags.AddError("Finished run's module source is nil.", "")
+		return diags
+	}
+	if finishedRun.ModuleVersion == nil {
+		diags.AddError("Finished run's module version is nil.", "")
+		return diags
+	}
+
+	// Return the output.
+	output.WorkspacePath = types.StringValue(finishedRun.WorkspacePath)
+	output.ModuleSource = types.StringValue(*finishedRun.ModuleSource)
+	output.ModuleVersion = types.StringValue(*finishedRun.ModuleVersion)
+	output.Variables = input.model.Variables // Cannot get variables back from a workspace or run, so pass them through.
+	return nil
+}
+
+func (t *workspaceDeployedModuleResource) waitForJobCompletion(ctx context.Context, jobID *string) error {
+	if jobID == nil {
+		return fmt.Errorf("nil job ID")
+	}
+
+	// Poll until job has finished.
+	for {
+
+		job, err := t.client.Job.GetJob(ctx, &sdktypes.GetJobInput{
+			ID: *jobID,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get job ID %s", *jobID)
+		}
+
+		if job.Status == "finished" {
+			return nil
+		}
+
+		time.Sleep(jobCompletionPollInterval)
+	}
+
+}
+
+// getCurrentDeployment returns a WorkspaceDeployedModuleModel reflecting what is currently deployed.
+func (t *workspaceDeployedModuleResource) getCurrentDeployment(ctx context.Context,
+	tfState WorkspaceDeployedModuleModel, target *WorkspaceDeployedModuleModel) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	// Get latest run on the target workspace.
+	wsPath := tfState.WorkspacePath.ValueString()
+	ws, err := t.client.Workspaces.GetWorkspace(ctx, &sdktypes.GetWorkspaceInput{
+		Path: &wsPath,
+	})
+	if err != nil {
+		diags.AddError(fmt.Sprintf("Failed to get specified workspace by path: %s", wsPath), err.Error())
+		return diags
+	}
+	latestRun, err := t.client.Run.GetRun(ctx, &sdktypes.GetRunInput{
+		ID: ws.CurrentStateVersion.RunID,
+	})
+	if err != nil {
+		diags.AddError("Failed to get latest run", err.Error())
+		return diags
+	}
+
+	// Make sure the module source and module version are not nil.
+	if latestRun.ModuleSource == nil {
```

Comment:
The `ModuleSource` and version may be nil if the workspace was applied using a configuration version or if the state was manually updated.